Implement authentication-oidc.login.specialgroup configuration option

danielcoelhocgu opened the pull request:
## References
* Fixes #11039
* Fixes #11570

## Description
Implements a new configuration option called `authentication-oidc.login.specialgroup` identical to the one provided by LDAP and Password plugins.

## Instructions for Reviewers
List of changes:
- Modifies the `getSpecialGroup` function in `OidcAuthenticationBean` class to return an EPerson Group considering the `authentication-oidc.login.specialgroup` configuration.
- Modifies the `authenticateWithOidc` function in `OidcAuthenticationBean` to set an attribute indicating OIDC Authentication when user is self registering. Otherwise the special group configured won't be available when user first logs in.

In order to test:
* Enable and configure OIDC Authentication
* Create an EPerson Group or choose an existing one and define any policy related to it (Ex: READ permission on a collection).
* Add the group name to the `authentication-oidc.login.specialgroup` option in the configuration file `local.cfg`
* Log in with OIDC and test the policy defined previously (Ex: try to access the restricted collection).

## Checklist

- [x] My PR is **created against the `main` branch** of code (unless it is a backport or is fixing an issue specific to an older branch).
- [x] My PR is **small in size** (e.g. less than 1,000 lines of code, not including comments & integration tests). Exceptions may be made if previously agreed upon.
- [x] My PR **passes Checkstyle** validation based on the [Code Style Guide](https://wiki.lyrasis.org/display/DSPACE/Code+Style+Guide).
- [x] My PR **includes Javadoc** for _all new (or modified) public methods and classes_. It also includes Javadoc for large or complex private methods.
- [x] My PR **passes all tests and includes new/updated Unit or Integration Tests** based on the [Code Testing Guide](https://wiki.lyrasis.org/display/DSPACE/Code+Testing+Guide).
- [x] My PR **includes details on how to test it**. I've provided clear instructions to reviewers on how to successfully test this fix or feature.
- [x] If my PR includes new libraries/dependencies (in any `pom.xml`), I've made sure their licenses align with the [DSpace BSD License](https://github.com/DSpace/DSpace/blob/main/LICENSE) based on the [Licensing of Contributions](https://wiki.lyrasis.org/display/DSPACE/Code+Contribution+Guidelines#CodeContributionGuidelines-LicensingofContributions) documentation.
- [x] If my PR modifies REST API endpoints, I've opened a separate [REST Contract](https://github.com/DSpace/RestContract/blob/main/README.md) PR related to this change.
- [x] If my PR includes new configurations, I've provided basic technical documentation in the PR itself.
- [x] If my PR fixes an issue ticket, I've [linked them together](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue).

danielcoelhocgu committed:
Implement authentication-oidc.login.specialgroup configuration option (#11039)

kanasznagyzoltan commented:
Hi [danielcoelhocgu](https://github.com/danielcoelhocgu), this looks like a very valuable enhancement! We are keen to test this feature in our environment.

We have a question about backward compatibility: Do you think it would be possible to apply this patch to older DSpace versions, specifically 8.2 or 9.1? We've briefly looked at the changeset and it seems like it might be compatible, but we wanted to ask for your expert opinion.

Thanks for your great work on this!

danielcoelhocgu committed:
Add authentication-oidc.login.specialgroup option to authentication-oidc.cfg file

danielcoelhocgu commented:
> @danielcoelhocgu Thanks for the PR! I haven’t had a chance to test it yet, but from a quick look at the code I think you’ll need to add a change to `dspace/config/modules/authentication-oidc.cfg` to declare the new `authentication-oidc.login.specialgroup` key, similar to how it’s done in other authentication methods. For example: https://github.com/DSpace/DSpace/blob/main/dspace/config/modules/authentication-ldap.cfg#L102-L107

Thank you @toniprieto. I've just added the key to the `authenticacion-oidc.cfg` config file and updated the PR.

toniprieto reviewed (approved):
Thank you, @danielcoelhocgu ! I was able to try it today, and I can confirm that it resolves both issues.

tdonohue reviewed (approved):
👍 Thanks @danielcoelhocgu !  This new configuration looks good to me. I'm not able to easily test it, but I trust the tests of others.  The code looks correct.
